### ISSUE TITLE
OverflowError: Range exceeds array bounds

### ISSUE THREAD

**Issue by lrq3000** [oracle/Skater#285]
I run into this problem when I try to use Skater (first time) on a catboost classifier.

Here is my code:
```
from skater.core.explanations import Interpretation
from skater.model import InMemoryModel

interpreter = Interpretation()

interpreter.load_data(X_train, feature_names=list(X_orig.columns))
im_model = InMemoryModel(model.predict_proba,
                         examples=X_test, 
                         #target_names=y_train.unique(),
                         #unique_values=y_train.unique()
                        )
plots = interpreter.feature_importance.plot_feature_importance(im_model, ascending=True)
```

Here is the output:
```
[2/7] features █████--------------- Time elapsed: 0 seconds
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-141-18afbd1218da> in <module>
----> 1 plots = interpreter.feature_importance.plot_feature_importance(im_model, ascending=True)

~\Anaconda3\lib\site-packages\skater\core\global_interpretation\feature_importance.py in plot_feature_importance(self, predict_fn, filter_classes, ascending, ax, progressbar)
    163             raise (MatplotlibDisplayError("Matplotlib unable to open display"))
    164 
--> 165         importances = self.feature_importance(predict_fn, filter_classes=filter_classes, progressbar=progressbar)
    166 
    167         if ax is None:

~\Anaconda3\lib\site-packages\skater\core\global_interpretation\feature_importance.py in feature_importance(self, model_instance, ascending, filter_classes, progressbar)
     85 
     86             if self.data_set.feature_info[feature_id]['numeric']:
---> 87                 samples = self.data_set.generate_column_sample(feature_id, n_samples=n, method='stratified')
     88             else:
     89                 samples = self.data_set.generate_column_sample(feature_id, n_samples=n, method='random-choice')

~\Anaconda3\lib\site-packages\skater\data\datamanager.py in generate_column_sample(self, feature_id, n_samples, method)
    407             return self._generate_column_sample_random_choice(feature_id, n_samples=n_samples)
    408         elif method == 'stratified':
--> 409             return self._generate_column_sample_stratified(feature_id, n_samples=n_samples)
    410         else:
    411             raise(NotImplementedError("Currenly we only support random-choice, stratified for "

~\Anaconda3\lib\site-packages\skater\data\datamanager.py in _generate_column_sample_stratified(self, feature_id, n_samples, n_bins)
    439         samples = []
    440         for window, n in zip(sample_windows, samples_per_bin):
--> 441             samples.append(np.random.uniform(window[0], window[1], size=int(n)).tolist())
    442 
    443         return np.array(flatten(samples))

mtrand.pyx in numpy.random.mtrand.RandomState.uniform()

OverflowError: Range exceeds valid bounds
```

**Comment by lrq3000** [comment]
After a bit more debugging, here is on what data it chokes:

```('myvar', nan, nan, 4)```

The two nan are window[0] and window[1]. So it means that my variable contains nans, and it's a float. I think this issue arises because Skater does not expect to see nan values in floats. How should I handle that for Skater to accept the input?

**Comment by lrq3000** [comment]
After more debugging, it's not about having nans in my variable. The issue comes from the following line:

```bins = list(np.percentile(self[feature_id], percentiles))```

Bins is always full of nan. But if I do manually the exact same calculation in my notebook, by copying over the percentiles generated by Skater, I get a correct result and no nans. I can't understand what's happening here, looks like an heisenbug.

**Comment by lrq3000** [comment]
Found the culprit: it's indeed because there are nan in the variable `X_train` as inputed in `interpreted.load_data()`. I thought it was using `X_test` as in `InMemoryModel()` since this is where the exception happened, but in fact it was still using `X_train` at this stage.

My quick and dirty fix: replace the line above (line 436 in `skater/data/datamanager.py`) by:

`bins = list(np.percentile(self[feature_id][~np.isnan(self[feature_id])], percentiles))`

This will fix the issue by ignoring nan. However it would be nice to allow for testing nan values too, since the missing values can be informative, and depending on the machine learning model used, they can be considered as a separate class on their own (such as with catboost or xgboost).

**Comment by lrq3000** [comment]
Note: this was all on Skater 1.0.2, the last version released on conda-forge, but there is now v1.1.2 by derickl:
https://anaconda.org/derickl/skater

I'll report here if I run into a similar issue.

**Comment by lrq3000** [comment]
With the latest release, I again get an error about nans in floats:

```
2019-08-19 10:42:16,328 - skater.core.explanations - WARNING - Progress bars slow down runs by 10-20%. For slightly 
faster runs, do progress_bar=False
[1/7] features ██------------------ Time elapsed: 2 seconds
2019-08-19 10:42:18,377 - skater.core.explanations - WARNING - Multiprocessing failed, going single process
[2/7] features █████--------------- Time elapsed: 2 seconds
---------------------------------------------------------------------------
RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: 
"""
Traceback (most recent call last):
  File "C:\Users\AI\Anaconda3\lib\site-packages\multiprocess\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "C:\Users\AI\Anaconda3\lib\site-packages\skater\core\global_interpretation\feature_importance.py", line 320, in compute_feature_importance
    strategy='uniform-over-similarity-ranks')
  File "C:\Users\AI\Anaconda3\lib\site-packages\skater\data\datamanager.py", line 447, in generate_column_sample
    return dm.generate_sample(*args, **kwargs)
  File "C:\Users\AI\Anaconda3\lib\site-packages\skater\data\datamanager.py", line 406, in generate_sample
    sorted_index = self._build_metastore()['sorted_index']
  File "C:\Users\AI\Anaconda3\lib\site-packages\skater\data\datamanager.py", line 265, in _build_metastore
    dists = cosine_distances(self.X, Y=medians).reshape(-1)
  File "C:\Users\AI\Anaconda3\lib\site-packages\sklearn\metrics\pairwise.py", line 681, in cosine_distances
    S = cosine_similarity(X, Y)
  File "C:\Users\AI\Anaconda3\lib\site-packages\sklearn\metrics\pairwise.py", line 1027, in cosine_similarity
    X, Y = check_pairwise_arrays(X, Y)
  File "C:\Users\AI\Anaconda3\lib\site-packages\sklearn\metrics\pairwise.py", line 112, in check_pairwise_arrays
    estimator=estimator)
  File "C:\Users\AI\Anaconda3\lib\site-packages\sklearn\utils\validation.py", line 542, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "C:\Users\AI\Anaconda3\lib\site-packages\sklearn\utils\validation.py", line 56, in _assert_all_finite
    raise ValueError(msg_err.format(type_err, X.dtype))
ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
"""

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
~\Anaconda3\lib\site-packages\skater\core\global_interpretation\feature_importance.py in feature_importance(self, model_instance, ascending, filter_classes, n_jobs, progressbar, n_samples, method, scorer_type, use_scaling)
    146             importance_dicts = []
--> 147             for importance in mapper(fi_func, arg_list):
    148                 importance_dicts.append(importance)

~\Anaconda3\lib\site-packages\multiprocess\pool.py in next(self, timeout)
    747             return value
--> 748         raise value
    749 

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-98-98f33cb11be1> in <module>
      1 # Plot barplot of features importance
----> 2 plots = interpreter.feature_importance.plot_feature_importance(im_model, ascending=False)

~\Anaconda3\lib\site-packages\skater\core\global_interpretation\feature_importance.py in plot_feature_importance(self, modelinstance, filter_classes, ascending, ax, progressbar, n_jobs, n_samples, method, scorer_type, use_scaling)
    258                                               method=method,
    259                                               scorer_type=scorer_type,
--> 260                                               use_scaling=use_scaling)
    261 
    262         if ax is None:

~\Anaconda3\lib\site-packages\skater\core\global_interpretation\feature_importance.py in feature_importance(self, model_instance, ascending, filter_classes, n_jobs, progressbar, n_samples, method, scorer_type, use_scaling)
    152             self.interpreter.logger.warn("Multiprocessing failed, going single process")
    153             importance_dicts = []
--> 154             for importance in map(fi_func, arg_list):
    155                 importance_dicts.append(importance)
    156                 if progressbar:

~\Anaconda3\lib\site-packages\skater\core\global_interpretation\feature_importance.py in compute_feature_importance(feature_id, input_data, estimator_fn, original_predictions, feature_info, feature_names, training_labels, method, scaled, scorer)
    318     if feature_info[feature_id]['numeric']:
    319         samples = copy_of_data_set.generate_column_sample(feature_id, n_samples=n,
--> 320                                                           strategy='uniform-over-similarity-ranks')
    321     else:
    322         samples = copy_of_data_set.generate_column_sample(feature_id, n_samples=n, strategy='random-choice')

~\Anaconda3\lib\site-packages\skater\data\datamanager.py in generate_column_sample(self, feature_id, *args, **kwargs)
    445                          feature_names=[feature_id],
    446                          index=self.index)
--> 447         return dm.generate_sample(*args, **kwargs)
    448 
    449 

~\Anaconda3\lib\site-packages\skater\data\datamanager.py in generate_sample(self, sample, include_y, strategy, n_samples, replace, bin_count)
    404 
    405         elif strategy == 'uniform-over-similarity-ranks':
--> 406             sorted_index = self._build_metastore()['sorted_index']
    407             range_of_indices = list(range(len(sorted_index)))
    408 

~\Anaconda3\lib\site-packages\skater\data\datamanager.py in _build_metastore(self)
    263 
    264         # how far each data point is from the global median
--> 265         dists = cosine_distances(self.X, Y=medians).reshape(-1)
    266 
    267         sorted_index = [self.index[i] for i in dists.argsort()]

~\Anaconda3\lib\site-packages\sklearn\metrics\pairwise.py in cosine_distances(X, Y)
    679     """
    680     # 1.0 - cosine_similarity(X, Y) without copy
--> 681     S = cosine_similarity(X, Y)
    682     S *= -1
    683     S += 1

~\Anaconda3\lib\site-packages\sklearn\metrics\pairwise.py in cosine_similarity(X, Y, dense_output)
   1025     # to avoid recursive import
   1026 
-> 1027     X, Y = check_pairwise_arrays(X, Y)
   1028 
   1029     X_normalized = normalize(X, copy=True)

~\Anaconda3\lib\site-packages\sklearn\metrics\pairwise.py in check_pairwise_arrays(X, Y, precomputed, dtype)
    110     else:
    111         X = check_array(X, accept_sparse='csr', dtype=dtype,
--> 112                         estimator=estimator)
    113         Y = check_array(Y, accept_sparse='csr', dtype=dtype,
    114                         estimator=estimator)

~\Anaconda3\lib\site-packages\sklearn\utils\validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
    540         if force_all_finite:
    541             _assert_all_finite(array,
--> 542                                allow_nan=force_all_finite == 'allow-nan')
    543 
    544     if ensure_min_samples > 0:

~\Anaconda3\lib\site-packages\sklearn\utils\validation.py in _assert_all_finite(X, allow_nan)
     54                 not allow_nan and not np.isfinite(X).all()):
     55             type_err = 'infinity' if allow_nan else 'NaN, infinity'
---> 56             raise ValueError(msg_err.format(type_err, X.dtype))
     57     # for object dtype data, we only check for NaNs (GH-13254)
     58     elif X.dtype == np.dtype('object') and not allow_nan:

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
```

The error is however more complex now as it is happening inside scikit-learn... As a quick workaround, I imputed my float columns with the median value, and I feed it to Skater, but not to my Catboost classifier which uses the original dataset with nan values, and it seems to work.

Here is my code for those with a similar issue:

```
# Simple custom imputer: replace floats nan values with the median
# This is necessary for Skater to work
X_train_imp = X_train.copy()
X_test_imp = X_test.copy()
for col in X_orig.columns:
    if X_orig[col].dtype == 'float':
        for x_imp in (X_train_imp, X_test_imp):
            x_imp.loc[x_imp[col].isnull(), col] = x_imp[col].median()
print('Imputing done')

from skater.core.explanations import Interpretation
from skater.model import InMemoryModel

interpreter = Interpretation(X_train_imp, feature_names=list(X_orig.columns))

im_model = InMemoryModel(model.predict_proba,
                         examples=X_test_imp,
                         target_names=y_train.unique()
                        )

# Plot barplot of features importance
plots = interpreter.feature_importance.plot_feature_importance(im_model, ascending=True)
```

I close this as there is no simple fix, scikit-learn needs to support nan values before Skater can (there is an `allow_nan` parameter inside the code but there is no public api documenting this feature so I guess it's still experimental or only for internal use).
